src/rust/aggregator/src/pipelines/order_history.rs

```diff
@@ -0,0 +1,324 @@
+        let last_indexed_timestamp =
+            sqlx::query_file!("sqlx_queries/order_history/get_last_indexed_timestamp.sql")
+                .fetch_optional(&mut transaction as &mut PgConnection)
+                .await
+                .map_err(to_pipeline_error)?;
+        if last_indexed_timestamp.is_none() {
+            return Ok(());
+        }
+        let last_indexed_timestamp = last_indexed_timestamp.unwrap().time;
+        let last_timestamp = sqlx::query_file!("sqlx_queries/order_history/get_last_time_indexed.sql")
+            .fetch_one(&mut transaction as &mut PgConnection)
+            .await
+            .map_err(to_pipeline_error)?
+            .time;
+        let last_timestamp = if let Some(e) = last_timestamp {
+            e
+        } else {
+            return Ok(());
+        };
+        let intervals = (last_timestamp - last_indexed_timestamp).num_minutes() as usize;
+        let mut timestamps = Vec::with_capacity(intervals);
+        for i in 0..intervals {
+            timestamps.push(last_indexed_timestamp + chrono::Duration::minutes(i as i64 + 1));
+        }
+        if timestamps.len() == 0 {
+            return Ok(());
+        }
+        let mut state = self.state.clone().unwrap_or_default();
+        for timestamp in &timestamps {
+            state = if state.orders.is_empty() && state.last_txn_indexed == BigDecimal::zero() {
+                self.get_state_from_timestamp_with_state(
+                    state.orders,
+                    &BigDecimal::zero(),
+                    timestamp,
+                    &mut transaction,
+                )
+                .await?
+            } else {
+                self.get_state_from_timestamp_with_state(
+                    state.orders,
+                    &state.last_txn_indexed,
+                    timestamp,
+                    &mut transaction,
+                )
+                .await?
+            };
+            for order in state.orders.values() {
+                sqlx::query_file!(
+                    "sqlx_queries/order_history/insert_order.sql",
+                    order.market_id,
+                    order.order_id,
+                    order.user,
+                    order.size,
+                    OrderDirection::from_bool_type(order.side, OrderType::Limit) as OrderDirection,
+                    order.price,
+                    timestamp,
+                )
+                .execute(&mut transaction as &mut PgConnection)
+                .await
+                .map_err(to_pipeline_error)?;
+            }
+        }
+        if let Some(last) = timestamps.last() {
+            sqlx::query_file!(
+                "sqlx_queries/order_history/update_last_indexed_timestamp.sql",
+                last,
+            )
+            .execute(&mut transaction as &mut PgConnection)
+            .await
+            .map_err(to_pipeline_error)?;
+        }
+        commit_transaction(transaction).await?;
+        self.state = Some(state);
```

Comment:
How does this handle hot restarts/upgrades?

Reply:
If the state is empty, it will calculate it from scratch using the events in the database (see [here](https://github.com/econia-labs/econia/pull/658/files/03a510d21ecbaff9dcc00fbf2e23f07779a2ce54#diff-e5ca5ff9be33fe28c5ce766144fe36c6af89372639b8002835f64e9be74e08e5R278-R286)).

This means that on aggregator startup, the first time this runs, it will take a little longer to calculate the state, but for following runs, it will use this precalculated state to generate the following states.